doc/syslog/python-logger-enhancement.md

```diff
@@ -0,0 +1,140 @@
+
+SONiC also provides a C/C++ logger implementation in `sonic-platform-common.common.logger.cpp`. This C/C++ logger implementation is also a wrapper of Linux standard `syslog` which is widely used by swss/syncd. It provides the ability to set log level on fly by starting a thread to listen to CONFIG DB LOGGER table change. SONiC infrastructure also provides the Python wrapper for `sonic-platform-common.common.logger.cpp` which is `swsscommon.Logger`. However, this logger implementation also has some drawbacks:
+
+
+1. `swsscommon.Logger` assumes redis DB is ready to connect. This is a valid assumption for swss/syncd. But it is not good for a Python logger implementation because some Python script may be called before redis server starting.
+2. `swsscommon.Logger` wraps Linux syslog which only support single log identifier for a daemon. 
+
+So, `swsscommon.Logger` is not an option too.
+
+This document describes a Python logger enhancement which allows user setting log level at run time.
+
+### Requirements
+
+- Allow user to change log level of `sonic_py_common.syslogger.SysLogger` at run time.
+- Logger instance shall use default log level if redis server is not up.
+
+Note: `sonic_py_common.logger.Logger` is not in the feature scope because it is going to be deprecated.
+
+### Architecture Design
+
+`swsscommon.Logger` depends on a thread to listen to CONFIG DB LOGGER table change. It refreshes log level for each logger instances once the thread detects a DB entry change. A thread is considered heavy in a python script, especially that there are many short and simple python scripts which also use logger. To keep python logger light weight, it uses a different design than `swsscommon.Logger`:
+
+- `SysLogger` class shall be changed to a singleton.
+- `SysLogger` instance shall load log level configuration from DB during initialization stage if DB configuration is available.
+- `SysLogger` instance shall save log level configuration to DB during initialization stage if DB configuration is not available.
+- Logger configuration shall be refreshed by CLI which send a SIGHUP signal to the daemon.
+
+![python-logger-enhancement](/doc/syslog/images/python_logger_enhancement.svg)
+
+### High-Level Design
+
+#### SysLogger class change
+
+- `SysLogger` class shall be changed to a singleton
+- A new argument `enable_runtime_config` shall be added to `SysLogger.__init__`. Default to False.
+- `SysLogger` instance shall load log level configuration from DB during initialization stage if `enable_runtime_config` is True and DB config is available.
+- `SysLogger` instance shall save log level configuration to DB during initialization stage if `enable_runtime_config` is True and DB configuration is not available.
+- A new class level method `refresh_config` shall be added to `SysLogger` class. This method handles load or save log level configuration logic.
```

Comment:
@Junchao-Mellanox can we rename the method to `update_log_level` ?

Reply:
sure